### ISSUE TITLE
[BUG]  train_text_to_image_lora.py not support Multi-nodes or Multi-gpus training.

### ISSUE THREAD

**Issue by WindVChen** [huggingface/diffusers#4046]
In [`train_text_to_image_lora.py`](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_lora.py), I notice that the LORA parameters are extracted into an `AttnProcsLayers` class:
```
518    lora_layers = AttnProcsLayers(unet.attn_processors)
```
And it is only the `lora_layers` that is wrapped by DistributedDataParallel in the following code:
```
670    lora_layers, optimizer, train_dataloader, lr_scheduler = accelerator.prepare(
            lora_layers, optimizer, train_dataloader, lr_scheduler
          )
```
In the training process, it seems that the `lora_layers` are not explicitly used but only the `unet` is used:
```
776    model_pred = unet(noisy_latents, timesteps, encoder_hidden_states).sample
```

My question is that when using Multi-GPUs or Multi-Machines, will the gradients be successfully averaged across all processes in the above way? 

It is true that in each process, the gradients will be backward to `unet.attn_processors`, and these gradients will be shared by `lora_layers`, so we can use `optimizer` to update the weights. However, since we actually use `unet.attn_processors` to do the forward operation, but not the wrapped `lora_layers`, can the gradients be correctly averaged? From [here](https://github.com/pytorch/pytorch/blob/e095716161f7133f2af79d69461b92aa8748fc9b/torch/nn/parallel/distributed.py#L1507), it seems that a wrapped module will have a different forward compared to its original forward operation.

I am not quite familiar with `torch.nn.parallel.DistributedDataParallel` wrapper, and I do worry about whether the current code in `train_text_to_image_lora.py` will lead to different LORA weights in different processes (if the gradients failed to broadcast among processes).

Hope to find some help here, thank you.

**Comment by WindVChen** [comment]
After carefully printing out the gradients and weights in different processes, it seems quite sure that the current LORA training script fails to be applied to Multi-nodes or Multi-GPUs training: The gradients failed to broadcast among the processes, which then lead to different LORA weights in different processes.

In the following, I give a printout when I start a 2-GPU task:
![image](https://github.com/huggingface/diffusers/assets/56162412/5638d089-f398-47ea-9ad5-69694ad9ce4f)

The first 3 number denotes `process/epoch/step`, and the following two tensors are the gradients and weights of `lora_layers.layers[31].to_out_lora.up` (A linear layer). It is obvious that the gradients are different in different processes. Thus, after three steps, it will turn to:
![image](https://github.com/huggingface/diffusers/assets/56162412/a6580490-9191-42f1-87dc-9927b81b5515)
Obviously, the weights of LORA are different in different processes.

I think it is a bug since there is no description of that in the script. People who use the script to train on multi-processes will finally and actually get the result on a single process.

**Comment by patrickvonplaten** [comment]
cc @williamberman @sayakpaul

**Comment by sayakpaul** [comment]
@WindVChen 

Thanks for elaborating on this! 

From gauging this briefly, it seems like passing `unet` to the prepare step rather than just the LoRA layers might just fix this. 

Could you confirm this once?

**Comment by WindVChen** [comment]
Hi @sayakpaul ,

Yes. Passing `unet` to the prepare step can quickly fix it. However, it also can bring some inconvenience: 1) The batch size has to be reduced by half, or the training will lead to out-of-memory; 2) The checkpoint stored intermediate is much larger (~3G compared to the original ~3M), as it needs to store the whole unet structure.

I wonder if there is a more elegant way to solve it without sacrificing memory and storage. I also tried wrapping each "loraAttnprocessor" manually, but found it required a lot of source code modification, so I gave up :cry:.

**Comment by sayakpaul** [comment]
Thanks for sharing!

> I also tried wrapping each "loraAttnprocessor" manually, but found it required a lot of source code modification, so I gave up 

Could you expand a bit more on this?

**Comment by WindVChen** [comment]
Yes.

It seems that the problem above happened just because the training script used the **unwrapped unet** to do the forward, not the **wrapped lora_layers**, so my previous idea is to wrap every `loraAttnprocessor` in the `unet` one by one (directly use `accelerate.prepare`)  and then replace the original **unwrapped loraAttnprocessors** in `unet` with these **wrapped loraAttnprocessors**. In this way, maybe we can expect that in the training phase, the gradients can broadcast among different processes.

However, since I'm not quite familiar with `torch.nn.parallel.DistributedDataParallel`, I'm not sure whether this solution can work.

**Comment by muellerzr** [comment]
When using Accelerate any models that get gradient/weight updates should be passed to `.prepare`. Do note that gradient accumulation currently won't work with multiple models, we're adding that very soon, but hope that helps.

**Comment by WindVChen** [comment]
Hi @muellerzr ,

Could you expand a bit more about "gradient accumulation currently won't work with multiple models"? Or are there any blogs or issues related to this? Because I am going to use gradient accumulation for multiple models.:sweat_smile:

**Comment by muellerzr** [comment]
No one has quite pointed out this "issue" yet actually, once the PR is opened (probably today or tomorrow) I'll link to it but basically multiple forward passes followed by multiple backward passes (for each model's loss, for instance) leads to some headaches in torch distributed.

**Comment by WindVChen** [comment]
OK, thanks.

Hope to be sure. So does that mean that a GAN training script like the one below will cause some BUGs?
```python
"Suppose gradient_accumulation is set to 2"
optimizer_gen = optim(generator.parameters())
optimizer_disc = optim(discriminator.parameters())
with accelerator.accumulate(generator):
        outputs = optimizer_gen(input)
        loss = loss_func(outputs)
        loss.backward()
        optimizer_gen.step()
        optimizer_gen.zero_grad()

        outputs = optimizer_disc(input)
        loss = loss_func(outputs)
        loss.backward()
        optimizer_disc.step()
        optimizer_disc.zero_grad()
```

**Comment by muellerzr** [comment]
Yes because technically the `discriminator` should also be under `accumulate`. So the discriminator might still be getting updated each step. (Though again, working on support there)

**Comment by WindVChen** [comment]
Ah, I see. Maybe I can double-wrap the pipeline for a temporary fix before the PR? Like this:
```python
with accelerator.accumulate(generator):
    with accelerator.accumulate(discriminator):
        outputs = optimizer_gen(input)
        ...
```

**Comment by muellerzr** [comment]
Double wrapping also does not work, hence the need for a more complex solution. (see discussion here: https://github.com/huggingface/accelerate/pull/1708). So just wait a day or two :)

**Comment by WindVChen** [comment]
OK, thanks a lot. Look forward to the solution. :blush:

**Comment by eliphatfs** [comment]
I have a solution that is not elegant but works: wrapping around it with another network.
```python
class SuperNet(torch.nn.ModuleDict):
    def forward(self, text_encoder, unet, batch, class_labels, noisy_model_input, timesteps):
        # Get the text embedding for conditioning
        encoder_hidden_states = encode_prompt(
            text_encoder,
            batch["input_ids"],
            None
        )
        # Predict the noise residual
        return unet(
            noisy_model_input, timesteps, encoder_hidden_states, class_labels=class_labels
        ).sample
```
Only use this module with accelerate and in the training loop instead of the original code call this module.

This originates as a solution to training text encoder and unet simultaneously for dreambooth. You could look here: https://github.com/huggingface/accelerate/issues/668#issuecomment-1614043548

**Comment by eliphatfs** [comment]
I have a solution that is not elegant but works: wrapping around it with another network.
```python
class SuperNet(torch.nn.ModuleDict):
    def forward(self, text_encoder, unet, batch, class_labels, noisy_model_input, timesteps):
        # Get the text embedding for conditioning
        encoder_hidden_states = encode_prompt(
            text_encoder,
            batch["input_ids"],
            None
        )
        # Predict the noise residual
        return unet(
            noisy_model_input, timesteps, encoder_hidden_states, class_labels=class_labels
        ).sample
```
Only use this module with accelerate and in the training loop instead of the original code call this module.

This originates as a solution to training text encoder and unet simultaneously for dreambooth. You could look here: https://github.com/huggingface/accelerate/issues/668#issuecomment-1614043548

**Comment by WindVChen** [comment]
Hi @eliphatfs ,

Thanks for sharing. Just want to confirm that the code given is actually solving the gradient accumulation problem of multiple models when using DDP, not the problem of applying multiple GPUs/nodes to the LORA code, is it? (There are two problems mentioned in this issue 😂 )

**Comment by eliphatfs** [comment]
> Hi @eliphatfs ,
> 
> Thanks for sharing. Just want to confirm that the code given is actually solving the gradient accumulation problem of multiple models when using DDP, not the problem of applying multiple GPUs/nodes to the LORA code, is it? (There are two problems mentioned in this issue 😂 )

You can do similar things for LoRA. This is from one of my custom pipelines, but can be similarly changed.
```python
class SuperNet(AttnProcsLayers):
    def forward(self, image_encoder, unet, image, noisy_model_input, timesteps):
        encoded = image_encoder(image)
        return unet(
            noisy_model_input,
            timesteps,
            encoder_hidden_states=encoded.last_hidden_state,
            class_labels=noise_image_embeddings(encoded.image_embeds, 0)
        ).sample
```
I think the thing is to make sure that
1) No mutable parameters are accessed outside the DDP wrapper forward and all of them are parameters of that wrapped module.
2) Only 1 DDP wrapper exists for each process group (or accelerator instance).

**Comment by WindVChen** [comment]
Thanks for the further description. 

But from my understanding, this may still fail to solve the problem listed [here](https://github.com/huggingface/diffusers/issues/4046#issuecomment-1632392537)? Have any thoughts on that problem? Or am I missing something?

**Comment by eliphatfs** [comment]
You are not passing the complete unet into accelerate, it is passed as an argument at forward time, so the parameters will not be stored in the checkpoint. You have to make sure all optimzied parameters are registered in the SuperNet of course.

**Comment by eliphatfs** [comment]
Anyone can verify if this has been fixed in accelerate?

**Comment by thuanz123** [comment]
Hi so the quick fix for this is including the unet as well as the lora_layers to the prepare step right ?

**Comment by sayakpaul** [comment]
Cc: @muellerzr

**Comment by ai1361720220000** [comment]
> This should be fixed by passing multiple models with `accelerator.accumulate`, yes @hkunzhe

Hi @muellerzr ,
What should be done when it comes to accelerator.prepare()? Does multi models are all put into "prepare" function or only one of them is ok?like accelerator.prepare(model1, model2, optimizer,train_dataloader, lr_scheduler) or accelerator.prepare(model1, optimizer,train_dataloader, lr_scheduler) , accelerator.prepare(model2, optimizer,train_dataloader, lr_scheduler).

**Comment by muellerzr** [comment]
All of them should be put into `prepare`, specifically all the ones that expect to have their gradients updated. Those same ones should then also be passed to `accumulate`.

You can send both into `prepare` at the same time.

**Comment by snowflakewang** [comment]
> I have a solution that is not elegant but works: wrapping around it with another network.
> 
> ```python
> class SuperNet(torch.nn.ModuleDict):
>     def forward(self, text_encoder, unet, batch, class_labels, noisy_model_input, timesteps):
>         # Get the text embedding for conditioning
>         encoder_hidden_states = encode_prompt(
>             text_encoder,
>             batch["input_ids"],
>             None
>         )
>         # Predict the noise residual
>         return unet(
>             noisy_model_input, timesteps, encoder_hidden_states, class_labels=class_labels
>         ).sample
> ```
> 
> Only use this module with accelerate and in the training loop instead of the original code call this module.
> 
> This originates as a solution to training text encoder and unet simultaneously for dreambooth. You could look here: [huggingface/accelerate#668 (comment)](https://github.com/huggingface/accelerate/issues/668#issuecomment-1614043548)

@eliphatfs Hello. Thank you for your good idea. I want to try your method. But I am not sure if I need to warp all models that are involved in the computing process, or just warp trainable models. Now I just warp trainable models as a unified class SuperNet. But after I pass SuperNet to accelerator.prepare, I printed SuperNet.parameters() on multiple devices and found them were different from each other, which meant that gradients were not synced across devices.
I wonder what cause such an issue.
Thank you! :)

**Comment by eliphatfs** [comment]
> > I have a solution that is not elegant but works: wrapping around it with another network.
> > ```python
> > class SuperNet(torch.nn.ModuleDict):
> >     def forward(self, text_encoder, unet, batch, class_labels, noisy_model_input, timesteps):
> >         # Get the text embedding for conditioning
> >         encoder_hidden_states = encode_prompt(
> >             text_encoder,
> >             batch["input_ids"],
> >             None
> >         )
> >         # Predict the noise residual
> >         return unet(
> >             noisy_model_input, timesteps, encoder_hidden_states, class_labels=class_labels
> >         ).sample
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Only use this module with accelerate and in the training loop instead of the original code call this module.
> > This originates as a solution to training text encoder and unet simultaneously for dreambooth. You could look here: [huggingface/accelerate#668 (comment)](https://github.com/huggingface/accelerate/issues/668#issuecomment-1614043548)
> 
> @eliphatfs Hello. Thank you for your good idea. I want to try your method. But I am not sure if I need to warp all models that are involved in the computing process, or just warp trainable models. Now I just warp trainable models as a unified class SuperNet. But after I pass SuperNet to accelerator.prepare, I printed the gradients of SuperNet.parameters() on multiple devices and found them were different from each other, which meant that gradients were not synced across devices. I wonder what cause such an issue. Thank you! :)

In my case only trainable models are needed to be wrapped. Only when the accumulation ends will the gradient be synced. During internal steps for accumulation backwards the gradients are not synced. I would usually check Adam momentums instead of the gradients, which shall always be in sync.

accelerate and diffusers has been updated by a lot since I worked on this, so I am not sure if it is due to a recent update that is breaking in your case.
